program/src/state/rules.rs

```diff
@@ -30,48 +29,54 @@ impl Rule {
         &self,
         accounts: &HashMap<Pubkey, &AccountInfo>,
         payload: &Payload,
-    ) -> ProgramResult {
+    ) -> (bool, RuleSetError) {
```

Comment:
wdyt about this:
```
impl Rule {
    pub fn validate(
        &self,
        accounts: &HashMap<Pubkey, &AccountInfo>,
        payload: &Payload,
    ) -> ProgramResult {
        let (status, rollup_err) = self.ll_validate(accounts, payload);

        if status {
            ProgramResult::Ok(())
        } else {
            ProgramResult::Err(rollup_err.into())
        }
    }

    fn ll_validate(
        &self,
        accounts: &HashMap<Pubkey, &AccountInfo>,
        payload: &Payload,
    ) -> (bool, RuleSetError) {
        match self {
            Rule::All { rules } => {
                msg!("Validating All");
                let mut last = self.to_error();
                for rule in rules {
                    last = rule.to_error();
                    let result = rule.ll_validate(accounts, payload);
                   ...
```
_(rest is same as what you have but recursive calls are to ll_validate())_

This would abstract the user from having to understand the tuple, but we still use it internally.

Reply:
Good call! I like that